### PR TITLE
Add a retry loop to UpdateStatus in NodeController

### PR DESCRIPTION
**Pull request by gmarek**

### CODE REVIEW

**Comment by vmarmol** [comment]
Assigning to @ddysher

File: pkg/cloudprovider/controller/nodecontroller.go [hunk 1]
```diff
@@ -446,54 +505,23 @@ func (nc *NodeController) MonitorNodeStatus() error {
...
+				node, err = nc.kubeClient.Nodes().Get(name)
+				if err != nil {
+					glog.Errorf("Failed while getting a Node to retry updating NodeStatus. Probably Node %s was deleted.", name)
+				}
+			} else {
+				break
 			}
 		}
+		if err != nil {
+			glog.Errorf("Update status  of Node %v from NodeController exceeds retry count. Skipping", node.Name)
```
Comment (davidopp) [inline pkg/cloudprovider/controller/nodecontroller.go]: should you "continue" after this line? if the update failed, I'm not sure you want to execute the pod eviction logic. probably better to err on the safe side.

File: pkg/cloudprovider/controller/nodecontroller.go [hunk 2]
```diff
@@ -446,54 +505,23 @@ func (nc *NodeController) MonitorNodeStatus() error {
...
-			}
-			_, err = nc.kubeClient.Nodes().Update(node)
+		for rep := 0; rep < nodeStatusUpdateRetry; rep++ {
+			err, gracePeriod, lastReadyCondition = nc.tryUpdateNodeStatus(node)
 			if err != nil {
-				glog.Errorf("error updating node %s: %v", node.Name, err)
+				name := node.Name
+				node, err = nc.kubeClient.Nodes().Get(name)
+				if err != nil {
+					glog.Errorf("Failed while getting a Node to retry updating NodeStatus. Probably Node %s was deleted.", name)
```
Comment (davidopp) [inline pkg/cloudprovider/controller/nodecontroller.go]: probably you should also break in this case. if the Get failed, calling tryUpdateNodeStatus() again with the old node seems pointless.

**Comment by davidopp** [comment]
Can you also please fix the bug that allows multiple simultaneous goroutines executing MonitorNodeStatus() at the same time (if one execution takes longer than 5 sec). This bug was already in the code before this PR but it probably makes sense to fix it in this PR.

File: pkg/cloudprovider/controller/nodecontroller.go (hunk 2 above)
Comment (gmarek) [inline pkg/cloudprovider/controller/nodecontroller.go]: Right. Thanks for the catch.

File: pkg/cloudprovider/controller/nodecontroller.go (hunk 1 above)
Comment (gmarek) [inline pkg/cloudprovider/controller/nodecontroller.go]: You're probably right. I was just keeping old semantics.

**Comment by gmarek** [comment]
@davidopp As for your last comment I don't think it's the case. NodeController Run method is called once (not in the loop), so only source of call to MonitorNodeStatus is from go util.Forever. After starting this goroutine Run method exits (this is the pattern in controller_manager).

util.Forever is effectively implemented as:
```
func Forever(f func(), period time.Duration) {
	for {
		func() {
			defer HandleCrash()
			f()
		}()
		time.Sleep(period)
	}
}
```

which means that 'period' is not an amount of time between consecutive beginnings of function runs, but amount of time which has to pass after the end of one call, before we begin another.

I'm also surprised that this is the way it works, but as a result I don't think there's a bug in place you mention.

**Comment by gmarek** [comment]
Friendly ping?

**Comment by davidopp** [comment]
LGTM

**Pull request merged**